enhancements/local-storage/automatic-detection.md

```diff
@@ -64,40 +65,299 @@ and we can use that information to sort devices into storage classes based on th
+    - ##### NOTE: This step is optional for users directly creating CRs. They might already know what CRs they need for step 2.
+2. PV Creation: After the user has decided how to create PVs, they can choose to create PVs with either the `LocalVolume` CR or the `LocalVolumeSet` CR.
+
+## Design Details for `Auto discovery of local devices`
+
+API scheme for `LocalVolumeDiscovery` CR:
+
+```go
+// DiscoveryPhase defines the observed phase of the discovery process
+type DiscoveryPhase string
+
+// Different phases of the discovery process
+const(
+  // DiscoveryComplete represents that the discovery process is complete
+  DiscoveryComplete DiscoveryPhase = "discovered"
+  // DiscoveryPending represents that the discovery process is in progress
+  DiscoveryPending DiscoveryPhase = "discovering"
+  // DiscoveryFailed represents that the discovery process has failed
+  DiscoveryFailed DiscoveryPhase = "error"
+)
+
+type LocalVolumeDiscoveryStatus struct {
+       // Phase represents the current phase of discovery process
+       // This is used by the OLM UI to provide status information
+       // to the user
+	Phase DiscoveryPhase `json:"phase,omitempty"`
+	// Conditions is a list of conditions and their status.
+	Conditions []operatorv1.OperatorCondition `json:"conditions,omitempty"`
+	// observedGeneration is the last generation change the operator has dealt with
+	// +optional
+	ObservedGeneration int64 `json:"observedGeneration,omitempty"`
+}
+
+type LocalVolumeDiscoverySpec struct {
+	// Nodes on which the automatic detection policies must run.
+	// +optional
+	NodeSelector *corev1.NodeSelector `json:"nodeSelector,omitempty"`
+	// If specified tolerations is the list of toleration that is passed to the
+	// LocalVolumeDiscovery Daemon
+	// +optional
+	Tolerations []corev1.Toleration `json:"tolerations,omitempty"`
+}
+
+type LocalVolumeDiscovery struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata"`
+	Spec              LocalVolumeDiscoverySpec   `json:"spec"`
+	Status            LocalVolumeDiscoveryStatus `json:"status,omitempty"`
+}
+
+type LocalVolumeDiscoveryList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata"`
+	Items           []LocalVolumeDiscovery `json:"items"`
+}
+```
+
+API scheme for `LocalVolumeDiscoveryResult` CR:
+```go
+// DeviceMechanicalProperty holds the device's mechanical spec. It can be rotational or nonRotational
+type DeviceMechanicalProperty string
+
+// The mechanical properties of the devices
+const (
+	// Rotational refers to magnetic disks
+	Rotational DeviceMechanicalProperty = "Rotational"
+	// NonRotational refers to ssds
+	NonRotational DeviceMechanicalProperty = "NonRotational"
+)
+
+// DeviceType represents the type of devices that would be discoverd.
+type DeviceType string
+
+// Devices types that should be discovered
+const (
+	// RawDisk represents a device-type of block disk
+	RawDisk DeviceType = "disk"
```

Comment:
Let's only discover the types of devices that are supported for local PVs. 
For example, [here](https://github.com/rook/rook/blob/f1b084daa99b811ebaa5c004015a4b832e801740/pkg/clusterd/disk.go#L38) are the types of disks that rook supports (definitions are [here](https://github.com/rook/rook/blob/f1b084daa99b811ebaa5c004015a4b832e801740/pkg/clusterd/disk.go#L38)). LSO might not support the same set as Rook, but it's a good place to start.

Reply:
Current LSO implementation (that is, `localVolume`) does not have any check for the supported device `types`.  It ignores devices with more than one mount points and devices with partitions. `LocalVolume` can also provision PV on partitions (/dev/sdc1, /dev/sdc2, etc) provided by the user in the CR. 

1. So currently supported types are `disk` and `partitions`? @gnufied  are there any other supported device types that we should discover?
2. Since LSO can provision PV on partitions, should we do discovery partitions as well ? (Earlier we decided not to discover partitions).

Reply:
In 4.7 we want to add new policy to the LocalVolumeSet for creating PVs on partitions. We will need the partition discovery at the same time, but we certainly don't need it for 4.6.

Reply:
partitions are required as per the current OCS story - @travisn @rohantmp  doesn't LocalVolumeSet create PV on partitions already?

Reply:
yup! LocalVolumeSet can use partitions as per the current implementation